### PR TITLE
formula_assertions: handle result `nil` case

### PR DESCRIPTION
**Pull request by chenrui333**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

seeing the failure in https://github.com/Homebrew/homebrew-core/actions/runs/10230575029/job/28305550081

```
    TypeError: Parameter 'result': Expected type Integer, got type NilClass
```

### CODE REVIEW

File: Library/Homebrew/formula_assertions.rb [hunk 1]
```diff
@@ -21,11 +21,11 @@ def assertions
     # Returns the output of running cmd and asserts the exit status.
     #
     # @api public
-    sig { params(cmd: T.any(Pathname, String), result: Integer).returns(String) }
+    sig { params(cmd: T.any(Pathname, String), result: T.nilable(Integer)).returns(String) }
     def shell_output(cmd, result = 0)
       ohai cmd
       output = `#{cmd}`
-      assert_equal result, $CHILD_STATUS.exitstatus
+      assert_equal result, $CHILD_STATUS.exitstatus unless result.nil?
```
Comment (carlocab) [inline Library/Homebrew/formula_assertions.rb:28]:
```suggestion
      assert_equal result, $CHILD_STATUS.exitstatus
```
I don't think you need this.

**Review by carlocab — DISMISSED** [review]
Weird formula.

**Comment by Bo98** [comment]
`shell_output` is probably the wrong tool for `debugbreak`. `nil` exit status means it's "crashing", but it doesn't indicate what of stop signal it's receiving - it could be a genuine progammer bug crash like a missing symbol.

In this case it's it's sending a SIGTRAP to signal the debugger to kick in. That test should probably be updated to run the command and then check something like `$CHILD_STATUS.stopsig == Signal.list["TRAP"]` instead

**Comment by chenrui333** [comment]
debugbreak is actually just one of the examples, there are some other formulae reporting nil status as well.

gopass is another one, https://github.com/Homebrew/homebrew-core/actions/runs/10230142627/job/28304642779

**Comment by carlocab** [comment]
> debugbreak is actually just one of the examples, there are some other formulae reporting nil status as well.

I don't think so:
```
❯ rg 'shell_output\(.*, nil\)'
d/debugbreak.rb
26:    assert_empty shell_output("./test", nil)
```

> gopass is another one, [Homebrew/homebrew-core/actions/runs/10230142627/job/28304642779](https://github.com/Homebrew/homebrew-core/actions/runs/10230142627/job/28304642779)

I don't see it: https://github.com/Homebrew/homebrew-core/blob/bf3b44194ef2048cc60cb2f8dd3af749c99e11d3/Formula/g/gopass.rb#L35-L58

**Comment by Bo98** [comment]
No parameter implies 0. So gopass is checking 0 not nil. The test is failing because the program is crashing - a genuine issue.

**Comment by chenrui333** [comment]
close in favor of https://github.com/Homebrew/homebrew-core/pull/179839

**Pull request closed without merging**